cmd/contour/servecontext.go

```diff
@@ -232,11 +233,33 @@ func (ctx *serveContext) tlsconfig() *tls.Config {
 		log.Fatalf("unable to append certificate in %s to CA pool", ctx.caFile)
 	}
 
+	// lazily load certificates and key at TLS handshake to ensure that latest
+	// certificates are used in case they have been rotated
 	return &tls.Config{
 		Certificates: []tls.Certificate{cert},
 		ClientAuth:   tls.RequireAndVerifyClientCert,
 		ClientCAs:    certPool,
 		Rand:         rand.Reader,
+		GetConfigForClient: func(*tls.ClientHelloInfo) (*tls.Config, error) {
+
+			cert, err := tls.LoadX509KeyPair(ctx.contourCert, ctx.contourKey)
+			check(err)
+
+			ca, err := ioutil.ReadFile(ctx.caFile)
+			check(err)
+
+			certPool := x509.NewCertPool()
+			if ok := certPool.AppendCertsFromPEM(ca); !ok {
+				log.Fatalf("unable to append certificate in %s to CA pool", ctx.caFile)
+			}
+
+			return &tls.Config{
+				Certificates: []tls.Certificate{cert},
+				ClientAuth:   tls.RequireAndVerifyClientCert,
+				ClientCAs:    certPool,
+				Rand:         rand.Reader,
+			}, nil
+		},
```

Comment:
I think this would be cleaner if you used an inner function to capture the common code:
```Go
func (ctx *serveContext) tlsconfig() *tls.Config {
	err := ctx.verifyTLSFlags()

	loadConfig := func() (*tls.Config, error) {
                // Load certificates, etc. Return the error, don't panic.
        }

        // Attempt to load certificates and key to catch configuration errors early.
        if _, err := loadConfig(); err != nil {
                // Fatal error here ...
        }

        return &tls.Config{
		ClientAuth: tls.RequireAndVerifyClientCert,
		Rand:       rand.Reader,
		GetConfigForClient: func(*tls.ClientHelloInfo) (*tls.Config, error) {
			return loadConfig(), nil
		},
	}
```

Reply:
This is good idea and I will do it, but just one question:

I recall that we shortly mentioned at the community meeting, that even in the case when certificates are be loaded at TLS connection establishment, we would keep the panic.  

We have these alternatives:

1. If the callback function returns error, the TLS implementation will send TLS alert to Envoy with message "internal error" and reject the connection establishment.
2. If the callback function returns nil, the TLS implementation will revert back to original `tls.Config`, causing Envoy eventually to reject the connection due to expired server certificate.
3. If the callback function panics, pod will be recreated and potentially might cause crash loop, similarily as the initial call to loadConfig() would do.

I was thinking that 3) panic might be easier for user to troubleshoot from these alternatives, but of course 1) or 2) could also have error print added to them...  What do you think?

Reply:
I agree with you that (3) is the best option. Make it a fatal error if `loadConfig` fails?

Reply:
It is now fixed accordingly.  

Small note though, since `golangci-lint` was giving following error

```
cmd/contour/servecontext.go:225:37: (*serveContext).tlsconfig$1 - result 1 (error) is always nil (unparam)
        loadConfig := func() (*tls.Config, error) {
                                           ^
```
I removed `error` return value from `loadConfig()` and documented in comments that all failures inside that function will cause immediate program exit.

Reply:
Ah, what I meant was to return an error from `loadConfig`, then do a fatal exit in the caller.

Reply:
Sorry my bad! Changed it now.

Reply:
The fatal exits were not tested previously either, but this change lowered coverage and got PR checks to fail. I can extract the inline function to test error separately, or alternatively use tricks from [here](https://stackoverflow.com/questions/26225513/how-to-test-os-exit-scenarios-in-go) to test functions that call `os.Exit()`.